lib/music_service/services/spotify.dart

```diff
@@ -0,0 +1,201 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:http/http.dart' as http;
+import 'package:song_tinder/music_service/music_service_interface.dart';
+import 'package:song_tinder/models/song_model.dart';
+import 'package:song_tinder/models/playlist_model.dart';
+import 'package:song_tinder/models/artist_model.dart';
+import 'package:song_tinder/models/album_model.dart';
+import 'package:url_launcher/url_launcher.dart';
+
+class Spotify extends MusicServiceInterface {
+  Spotify() {
+    _initAuth();
+  }
+
+  final client = http.Client();
+
+  late String accessToken;
+  static const String _clientId = ""; // Your Spotify app clientId;
+  static const _clientSecret = ""; // Your Spotify app clientSecret;
+
+  static const String _callbackServerURL = "http://localhost:3000";
+  final Uri _initAuthURL = Uri(
+      scheme: 'https',
+      host: 'accounts.spotify.com',
+      path: '/authorize',
+      queryParameters: {
+        'client_id': _clientId,
+        'redirect_uri': _callbackServerURL,
+        'response_type': 'code',
+        'scope':
+            'ugc-image-upload user-modify-playback-state user-read-playback-state user-read-currently-playing user-follow-modify user-follow-read user-read-recently-played user-read-playback-position user-top-read playlist-read-collaborative playlist-modify-public playlist-read-private playlist-modify-private app-remote-control streaming user-read-email user-read-private user-library-modify user-library-read'
+      });
+
+  
+  Future<http.Response> _sendRequest(String path, Map<String, String>? queryParameters, bool isPost) async {
+    http.Response response;
+    if (isPost) {
+      var uri = Uri.https('api.spotify.com', '/v1$path');
+      response = await client.post(uri, headers: {'Authorization': 'Bearer ' + accessToken}, body: queryParameters);
+    }
+    else {
+      var uri = Uri.https('api.spotify.com', "/v1$path", queryParameters);
+      print(uri.toString());
+      response = await client.get(uri, headers: {'Authorization': 'Bearer ' + accessToken});
+    }
+    if (response.statusCode == 401) {
+      _renewAuth();
+      return _sendRequest(path, queryParameters, isPost);
+    }
+    if (response.statusCode == 200) {
+      return response;
+    }
+    throw HttpException("Unexpected response status: ${response.statusCode}");
+  }
+
+  void _initAuth() async {
+    await launchUrl(
+        _initAuthURL); // open Spotify's authorization form on a web view
+
+    // start listener for callback response
+    await HttpServer.bind('0.0.0.0', 3000).then((listener) {
```

Comment:
What is this port for? Is this the callback for the auth? If so it would be great to give this constant a name